### PR TITLE
sweep: remove possible RBF when sweeping new inputs

### LINKED ISSUE #7972: [bug]: sweeper (bumpfee) creates invalid transaction (not paying higher fees when necessary)
### Background

The sweeper creates a combined sweep transaction, using another pending sweep transaction's input as an input, without raising the fee rate. The resulting combined transaction cannot be broadcast because of that.

### Your environment

* version of lnd: 0.16.4
* which operating system (uname -a on *Nix): Linux server 6.1.0-11-amd64 #<i></i>1 SMP PREEMPT_DYNAMIC Debian 6.1.38-4 (2023-08-08) x86_64 GNU/Linux
* version of btcd, bitcoind, or other backend: bitcoind v25

### Steps to reproduce

Run bumpfee for input A and 15 sat/vByte, resulting in a sweep transaction paying 15 sat/vByte.
Wait for the sweep transaction to broadcast.
Run bumpfee for input B and 15 sat/vByte before the sweep for input A is confirmed.

### Expected behaviour

Input A is not used as an input for the sweep transaction for input B.

### Actual behaviour

It is used, resulting in an error:
```
2023-09-10 11:59:22.323 [INF] SWPR: Creating sweep transaction xxx for 2 inputs (xxx:0 (CommitmentAnchor), xxx:0 (CommitmentAnchor)) using 253 sat/kw, tx_weight=774, tx_fee=0.00000195 BTC, parents_count=0, parents_fee=0 BTC, parents_weight=0
2023-09-10 11:59:22.333 [ERR] BTCN: Broadcast attempt failed: -26: insufficient fee, rejecting replacement xxx; new feerate 0.00001010 BTC/kvB <= old feerate 0.00004508 BTC/kvB
2023-09-10 11:59:22.337 [INF] LNWL: Inserting unconfirmed transaction xxx
2023-09-10 11:59:22.345 [INF] LNWL: xxx: broadcast failed because of: unable to replace transaction: -26: insufficient fee, rejecting replacement xxx; new feerate 0.00001010 BTC/kvB <= old feerate 0.00004508 BTC/kvB
2023-09-10 11:59:22.353 [ERR] SWPR: Publish sweep tx xxx got error: transaction rejected: output already spent
```

### PR DESCRIPTION
**Pull request by yyforyongyu**
Depends on some of the test framework from,
- #7965 
- #7824

Fixes #7972 and possible #7181

This PR removes a possible case that when sweeping new and old inputs, the new inputs may end up in a second transaction which accidentally attempts an RBF(and fails). It also refactors the sweeper and simplified the ticker logic, preparing for incoming refactors.

### CODE REVIEW

File: sweep/sweeper.go [hunk 1]
```diff
@@ -237,12 +237,21 @@ type UtxoSweeper struct {
 	wg   sync.WaitGroup
 }
 
+// feeDeterminor defines an alias to the function signature of
```
Comment (Roasbeef) [inline sweep/sweeper.go:241]: Should be `Determiner` to be "dictionary correct".

File: sweep/sweeper_test.go [hunk 2]
```diff
@@ -2239,3 +2240,120 @@ func TestSweeperShutdownHandling(t *testing.T) {
 	)
 	require.Error(t, err)
 }
+
+// TestFeeRateForPreference checks `feeRateForPreference` works as expected.
+func TestFeeRateForPreference(t *testing.T) {
```
Comment (Roasbeef) [inline sweep/sweeper_test.go:2137]: 👍

File: sweep/sweeper.go [hunk 3]
```diff
@@ -928,7 +928,6 @@ func (s *UtxoSweeper) clusterByLockTime(inputs pendingInputs) ([]inputCluster,
 	pendingInputs) {
 
 	locktimes := make(map[uint32]pendingInputs)
-	inputFeeRates := make(map[wire.OutPoint]chainfee.SatPerKWeight)
```
Comment (Roasbeef) [inline sweep/sweeper.go:915]: Solid change, maybe we added the fee rate attribute later on?

File: sweep/sweeper.go [hunk 4]
```diff
@@ -944,17 +944,25 @@ func (s *UtxoSweeper) clusterByLockTime(inputs pendingInputs) ([]inputCluster,
...
 		}
 
-		cluster[op] = input
-		locktimes[lt] = cluster
-
-		// We also get the preferred fee rate for this input.
+		// Get the fee rate based on the fee preference. If an error is
+		// returned, we'll skip sweeping this input for this round of
+		// cluster creation and retry it when we create the clusters
+		// from the pending inputs again.
```
Comment (Roasbeef) [inline sweep/sweeper.go:813]: In what case would an error be returned here? If an estimator for a neutrino client was down? But IIUC, that should just result in it using the cached or default value.

Comment (Roasbeef) [reply sweep/sweeper.go:813]: Either way, indeed a bug fix as before we would add the inputs, but then the fee rate would never be set due to the continue.

File: input/mocks.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import (
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/stretchr/testify/mock"
+)
+
+// MockInput implements the `Input` interface and is used by other packages for
+// mock testing.
+type MockInput struct {
+	mock.Mock
```
Comment (Roasbeef) [inline input/mocks.go:12]: Def gotta use these more. More ergonomic (and less deadlock prone!) that using a series of custom internal channels to send/recv args or return values.

File: lnwallet/chainfee/rates.go [hunk 6]
```diff
@@ -70,6 +70,11 @@ func (s SatPerKWeight) FeePerKVByte() SatPerKVByte {
 	return SatPerKVByte(s * blockchain.WitnessScaleFactor)
 }
 
+// FeePerVByte converts the current fee rate from sat/kw to sat/vb.
+func (s SatPerKWeight) FeePerVByte() SatPerVByte {
+	return SatPerVByte(s * blockchain.WitnessScaleFactor / 1000)
```
Comment (Roasbeef) [inline lnwallet/chainfee/rates.go:75]: I think we have a few other areas we do this custom calculation. Let's replace them with this instead? 

EDIT: did a basic sweep, and would seem not.

File: sweep/sweeper_test.go [hunk 7]
```diff
@@ -2340,6 +2340,7 @@ func TestFeeRateForPreference(t *testing.T) {
 		},
 	}
 
+	//nolint:paralleltest
```
Comment (Roasbeef) [inline sweep/sweeper_test.go:2234]: Once that other PR lands, we can get rid of this.

File: sweep/sweeper.go [hunk 8]
```diff
@@ -1651,6 +1582,85 @@ func (s *UtxoSweeper) ListSweeps() ([]chainhash.Hash, error) {
...
+
+	// Check to see if with this new input a sweep tx can be formed.
+	if err := s.scheduleSweep(bestHeight); err != nil {
+		log.Errorf("schedule sweep: %v", err)
+	}
+
+	log.Tracef("input %v scheduled", outpoint)
+}
+
+func (s *UtxoSweeper) handleExistingInput(input *sweepInputMessage,
```
Comment (Roasbeef) [inline sweep/sweeper.go:1513]: Missing `godoc` comment.

File: sweep/sweeper.go [hunk 9]
```diff
@@ -1019,7 +1020,7 @@ func (s *UtxoSweeper) scheduleSweep(currentHeight int32) error {
 
 	// Start sweep timer to create opportunity for more inputs to be added
 	// before a tx is constructed.
-	s.timer = s.cfg.NewBatchTimer()
+	s.timer = time.NewTicker(s.cfg.TickerDuration).C
```
Comment (Roasbeef) [inline sweep/sweeper.go]: Stop the prior ticker before? Would require changing the variable here or adding another one to be able to capture more than just the notification channel.

File: sweep/sweeper.go [hunk 10]
```diff
@@ -606,6 +603,12 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch) {
 		return
 	}
 
+	// Create a ticker based on the config duration.
+	ticker := time.NewTicker(s.cfg.TickerDuration)
```
Comment (Roasbeef) [inline sweep/sweeper.go:607]: +1 for storing the data in the [unified control flow](https://research.swtch.com/pcdata), instead of instances of independent control flow modifying the same piece of data

File: sweep/sweeper_test.go [hunk 11]
```diff
@@ -2436,3 +2430,107 @@ func TestClusterByLockTime(t *testing.T) {
...
+		Output: &wire.TxOut{Value: 100_000},
+	})
+
+	// Create a new and a retried input.
+	newInput := &pendingInput{
+		Input:  testInput,
+		params: param,
+	}
+	oldInput := &pendingInput{
+		Input:           testInput,
```
Comment (Roasbeef) [inline sweep/sweeper_test.go:2460]: Should these be spending the same input? Or the detail doesn't matter for this test context?

**Review by Roasbeef — COMMENTED** [review]
Solid set of refactoring and additional tests! I need to sit a bit more with the RBF change. IIUC, this shouldn't affect our sweeping unless the new sweep actually violates a rule of RBF. One example would be using the same transaction, but adding an a _new unconfirmed_ input.

File: sweep/sweeper_test.go (hunk 11 above)
Comment (yyforyongyu) [reply sweep/sweeper_test.go:2465]: correct it doesn't matter - update the comments to make it more clear

File: sweep/sweeper.go (hunk 10 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:607]: oh nice blog post!

File: sweep/sweeper.go (hunk 9 above)
Comment (yyforyongyu) [reply sweep/sweeper.go]: Could do. Since this is removed in the next commit, so I guess it's fine to leave it as-is?

File: sweep/sweeper.go (hunk 8 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:1514]: updated!

File: sweep/sweeper.go (hunk 1 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:241]: fixed!

File: sweep/sweeper.go (hunk 4 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:813]: Both `bitcoind` and `btcd` estimators won't return an error from calling `feeEstimator.EstimateFeePerKW`, so the only error would be from calling `feeEstimator.EstimateFeePerKW` in `WebAPIEstimator` with a conf target of 0. Aside from that, we could get an error if the fee rate is too low(below relay fee), or the fee preference is misused, like setting both conf target and fee rate - but these should all be handled internally atm.

File: input/mocks.go (hunk 5 above)
Comment (yyforyongyu) [reply input/mocks.go:12]: yeah I use it a lot, especially in the [new payment lifecycle tests](https://github.com/lightningnetwork/lnd/pull/7503/commits/f3d435b7cc54064bdd5ced15e9da5b12b5968a60#diff-4963b0f520937f7e5a4aa825c3541d17af3514a2c7d8f94ce83ccbe4308b0739R692) where the control flow is easily handled.

File: sweep/sweeper.go (hunk 3 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:915]: did a git blame and it seems it was added before

File: lnwallet/chainfee/rates.go (hunk 6 above)
Comment (yyforyongyu) [reply lnwallet/chainfee/rates.go:75]: cool found a few other places and replaced.

File: sweep/sweeper.go [hunk 12]
```diff
@@ -687,22 +687,57 @@ func (s *UtxoSweeper) removeExclusiveGroup(group uint64) {
...
-	// funding of a channel.
+	// Execute the sweep within a coin select lock. Otherwise the coins
+	// that we are going to spend may be selected for other transactions
+	// like funding of a channel.
 	return s.cfg.Wallet.WithCoinSelectLock(func() error {
-		// Examine pending inputs and try to construct
-		// lists of inputs.
-		inputLists, err := s.getInputLists(cluster, currentHeight)
+		// Examine pending inputs and try to construct lists of inputs.
+		allSets, newSets, err := s.getInputLists(cluster, currentHeight)
```
Comment (ziggie1984) [inline sweep/sweeper.go:695]: Hmm I am a bit undecided whether it's worth it introducing this short term fix in this way, why:

Let's pretend we do not have a node restart yet (because this marks every not swept input as new again), we are currently not really verifying the RBF constraints, this means we could easily bundle new sweeps with retried one(already in mempool) and not fullfill the BIP125 requirements. 

What I propose instead would be:

1. Exclude the AnchorOuputs in a different input set (the anchors sweeps which are swept at min realy fee) => because third party sweeps which bundle a significant amount of anchors can brings us in negative-yielding territory very fast, therefore its important to exclude them into a different sweep set.
2. I would switch the order meaning that I would first sweep the new inputs first. Because we do not verify the RBF conditions yet, we should not risk bundle an input with an already published one, especially until 1) is not fixed, but also imagine fees go done, we would bundle the new inputs with retired one (published when fees were higher) reducing the successful broadcasting of the tx to almost zero.
3. After every restart of lnd we should definitely check for all inputs and whether they are already in our mempool or even spent (using the `gettxout` rpc call) because rescanning can take a lot of time I experienced on my system. When we have a clear view which Inputs are already in our mempool, we still have a clear view between new inputs and retried ones. 

So to sum it up, I think we need to take care of 1) first. Until we have a clear picture of whats happening in our mempool and how many inputs are already sitting there at which feerate, we should broadcast the newsets first (even don't mix retired inputs with new inputs yet).

**Review by ziggie1984 — COMMENTED** [review]
I only realized after reviewing the first part, that it's another PR 😅, so left nits I did not seeing when I reviewed the base PR for it.

In terms of the new PR, I really like the tests and the refactoring. What I am not so sure about is the fix of the RBF case,  because I think deteriorates the sweeper situation mainly because most node runner have those min-relay-anchors sitting around and marking almost all sweeps invalid as soon as a third party bundled a big tx with a lot of anchors (increasing the RBF condition). So in the old case, we would at least publish the newSet as well, having a change for the new inputs to get broadcasted. LMKWYT ?

File: sweep/sweeper.go [hunk 13]
```diff
@@ -1651,6 +1582,85 @@ func (s *UtxoSweeper) ListSweeps() ([]chainhash.Hash, error) {
...
+	log.Tracef("input %v scheduled", outpoint)
+}
+
+// handleExistingInput processes an input that is already known to the sweeper.
+// It will overwrite the params of the old input with the new ones.
+func (s *UtxoSweeper) handleExistingInput(input *sweepInputMessage,
+	oldInput *pendingInput) {
+
+	// Before updating the input details, check if an exclusive group was
+	// set, and if so, assume this input as finalized and remove all other
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper.go:1518]: very nice comments!

File: sweep/sweeper_test.go [hunk 14]
```diff
@@ -2239,3 +2240,120 @@ func TestSweeperShutdownHandling(t *testing.T) {
...
+
+	// errFeeFunc is a mock over DetermineFeePerKw that always return the
+	// above dummy error.
+	errFeeFunc := func(_ chainfee.Estimator, _ FeePreference) (
+		chainfee.SatPerKWeight, error) {
+
+		return 0, dummyErr
+	}
+
+	// Set the relay fee rate to be 1.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper_test.go:2153]: fixed

File: sweep/sweeper_test.go [hunk 15]
```diff
@@ -2239,3 +2240,120 @@ func TestSweeperShutdownHandling(t *testing.T) {
...
+	}{
+		{
+			// When the fee preference is empty, we should see an
+			// error.
+			name:        "empty fee preference",
+			feePref:     FeePreference{},
+			expectedErr: ErrNoFeePreference,
+		},
+		{
+			// When an error is returned from the fee determinor,
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper_test.go:2201]: fixed

File: sweep/sweeper.go [hunk 16]
```diff
@@ -928,7 +928,6 @@ func (s *UtxoSweeper) clusterByLockTime(inputs pendingInputs) ([]inputCluster,
 	pendingInputs) {
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper.go:791]: too bad it's gone🙈

File: sweep/sweeper.go [hunk 17]
```diff
@@ -1661,6 +1581,82 @@ func (s *UtxoSweeper) handleExistingInput(input *sweepInputMessage,
...
+	if err != nil {
+		log.Errorf("cannot determine if tx %v is ours: %v",
+			spendHash, err)
+		return
+	}
+
+	// If this isn't our transaction, it means someone else swept outputs
+	// that we were attempting to sweep. This can happen for anchor outputs
+	// as well as justice transactions. In this case, we'll notify the
+	// wallet to remove any spends that a descent from this output.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper.go:1561]: fixed

File: sweep/sweeper.go [hunk 18]
```diff
@@ -1661,6 +1581,82 @@ func (s *UtxoSweeper) handleExistingInput(input *sweepInputMessage,
...
+	if !isOurTx {
+		err := s.removeLastSweepDescendants(spend.SpendingTx)
+		if err != nil {
+			log.Warnf("unable to remove descendant transactions "+
+				"due to tx %v: ", spendHash)
+		}
+
+		log.Debugf("Detected spend related to in flight inputs "+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper.go:1569]: updated

File: sweep/sweeper.go [hunk 19]
```diff
@@ -956,25 +955,24 @@ func (s *UtxoSweeper) clusterByLockTime(inputs pendingInputs) ([]inputCluster,
 		}
 
 		input.lastFeeRate = feeRate
-		inputFeeRates[op] = feeRate
 	}
 
 	// We'll then determine the sweep fee rate for each set of inputs by
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper.go:831]: the plan here is to remove the locktime field - we track the spending condition outside the sweeper, the caller won't send a sweep request unless the input has been mature, so this locktime field has no utility here.

File: sweep/sweeper.go (hunk 8 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:1514]: > Or is there another usecase already ?

Yes! If you look at how we update the params via `BumpFee` and how we handle existing inputs, it's clear the logic can be merged, something to add maybe in the following PR?

File: sweep/sweeper.go [hunk 20]
```diff
@@ -606,6 +603,12 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch) {
 		return
 	}
 
+	// Create a ticker based on the config duration.
+	ticker := time.NewTicker(s.cfg.TickerDuration)
+	defer ticker.Stop()
+
+	log.Debugf("Sweep ticker started")
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply sweep/sweeper.go:611]: This reminds me of our previous discussion - we need to clearly define what `force` means here. Atm it only means to sweep it even it's uneconomical, and it does make sense to also kick off the sweeper immediately. So in my mind there are two places to bypass the ticker,
- when `force` is set, and,
- when a user calls `sweepinputs` - a new RPC method that allow users to kick off the sweep manually(TBD)

File: sweep/sweeper.go (hunk 12 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:695]: So the general direction of refactoring sweeper is to remove the contextual info such as anchors or second level HTLCs, and focusing only on creating/publishing transactions **economically**. And all the contextual info should be abstracted and specified via the `Params` field. I don't know this is possible or not, still digging. That being said, the sweeper shouldn't care whether this is an anchor input or other kinds, you just tell it how you wanna it being swept via the `Params` field.

As for the order of new inputs vs all inputs, tbh I'm not sure here - my thought is we should try to sweep all inputs first, and if there's an error, we then sweep the new inputs, otherwise the old inputs would just be sitting there. There's also a chance that the RBF condition will be met when we sweep all inputs, for instance the mempool conditions changed and the fee estimator now gives a higher fee rate - of course we don't wanna relay on this uncontrolled behavior, which is something we'll fix in the following PR to handle the RBF case properly.

**Comment by yyforyongyu** [comment]
> What I am not so sure about is the fix of the RBF case, because I think it deteriorates the sweeper situation mainly because most node runners have those min-relay-anchors sitting around and marking almost all sweeps invalid as soon as a third party bundled a big tx with a lot of anchors (increasing the RBF fee condition). So in the old case, we would at least publish the newSet as well, having a change for the new inputs to get broadcasted. LMKWYT ?

There are two RBF happening at the moment,
1. when creating sweeping txes, if there are mixed inputs, at least two txes are created - `txAll` for all inputs, `txNew` for new inputs. And the possible result,
    - `txAll` gets in mempool and be replaced by `txNew`, or the other way around where `txNew` gets replace. Tho unwanted, this is at least a good case as there's no broadcasting errors.
    - `txAll` gets in mempool and `txNew` gets a broadcast error, or the other way around, and we don't know which case will happen because we can't guarantee the tx propagation order.
2. when sweeping all inputs, retried inputs are already in mempool, thus creating an RBF case which may or may not be satisfied.

This PR aims to fix case 1 so we don't have this RBF condition - we can try broadcast `txAll` first, if it fails, we then broadcast `txNew`.

File: sweep/sweeper.go (hunk 8 above)
Comment (ziggie1984) [reply sweep/sweeper.go:1519]: Great, yes let's do it in the follow-up PR

File: sweep/sweeper.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (ziggie1984) [reply sweep/sweeper.go:611]: jep, that's also how I see it. You are right force currently only allows for negative yielding (which for me signals that it's an important sweep).  So are we adding the by-passing part for the `force` also in a follow-up ?

File: sweep/sweeper.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (ziggie1984) [reply sweep/sweeper.go]: understand, tho you have to be careful, because send-level htlcs commit to the exact locktime whereas there are situations where other outputs commit to a `greater than` locktime. So both will be offered to the sweeper when they are ready to be spent, tho you have to be careful because you cannot mix them in some situations.

File: sweep/sweeper.go (hunk 12 above)
Comment (ziggie1984) [reply sweep/sweeper.go:695]: > As for the order of new inputs vs all inputs, tbh I'm not sure here - my thought is we should try to sweep all inputs first, and if there's an error, we then sweep the new inputs, otherwise the old inputs would just be sitting there.

That sounds good to me as well, but with the current change we would not do this or, at least not until all retried inputs reached their maximum retry attempt(default=15)? By that time we could in the worst case have to wait quite a bit until the conflicting inputs are gone ? The code before this PR would at least try them as well, so my concern is that we are not temporarily (until follow-up PR's come in)  improving the sweeper code with this change or but I might miss something here ?

File: sweep/sweeper.go [hunk 21]
```diff
@@ -1651,6 +1582,94 @@ func (s *UtxoSweeper) ListSweeps() ([]chainhash.Hash, error) {
...
+		log.Debugf("Already pending input %v received", outpoint)
+
+		s.handleExistingInput(input, pendInput)
+
+		return
+	}
+
+	// Create a new pendingInput and initialize the listeners slice with
+	// the passed in result channel. If this input is offered for sweep
+	// again, the result channel will be appended to this slice.
```
Comment (GeorgeTsagk) [inline sweep/sweeper.go:1491]: the reason you append to this slice is to notify all of the requesters?

File: sweep/sweeper.go [hunk 22]
```diff
@@ -1670,6 +1590,82 @@ func (s *UtxoSweeper) handleExistingInput(input *sweepInputMessage,
...
+	// as well as justice transactions. In this case, we'll notify the
+	// wallet to remove any spends that descent from this output.
+	if !isOurTx {
+		err := s.removeLastSweepDescendants(spend.SpendingTx)
+		if err != nil {
+			log.Warnf("unable to remove descendant transactions "+
+				"due to tx %v: ", spendHash)
+		}
+
+		log.Debugf("Detected third party spend related to in flight "+
```
Comment (GeorgeTsagk) [inline sweep/sweeper.go:1583]: should this be raised to a more visible log level, e.g `Warnf` ?

File: sweep/sweeper.go (hunk 12 above)
Comment (GeorgeTsagk) [reply sweep/sweeper.go:695]: About the anchor inputs: I also think that placing anchors in the same tx as other sweeps may introduce extra complexity but also higher fees overall. Maybe they could have their own `ExclusiveGroup` which is anchors-only :+1: .

The rationale for this would be that anchors are lower-priority sweeps of inferior significance. If there is an on-going sweep attempt for inputs of high significance, adding anchor inputs is gonna increase the total weight of the tx which is meant to be swept fast (maybe because of other inputs' timeouts).

File: sweep/sweeper.go [hunk 23]
```diff
@@ -687,22 +687,62 @@ func (s *UtxoSweeper) removeExclusiveGroup(group uint64) {
...
-	// funding of a channel.
+	// Execute the sweep within a coin select lock. Otherwise the coins
+	// that we are going to spend may be selected for other transactions
+	// like funding of a channel.
 	return s.cfg.Wallet.WithCoinSelectLock(func() error {
-		// Examine pending inputs and try to construct
-		// lists of inputs.
-		inputLists, err := s.getInputLists(cluster, currentHeight)
+		// Examine pending inputs and try to construct lists of inputs.
+		allSets, newSets, err := s.getInputLists(cluster, currentHeight)
```
Comment (GeorgeTsagk) [inline sweep/sweeper.go:695]: seems like `allSets` only returns sets of previously-published sweeps. Why not reflect that on the name, e.g `existingSets` ?

Comment (yyforyongyu) [reply sweep/sweeper.go:695]: `allSet` uses all the inputs to make the sweeps tho.

File: sweep/sweeper.go [hunk 24]
```diff
@@ -859,22 +687,62 @@ func (s *UtxoSweeper) removeExclusiveGroup(group uint64) {
...
-			err := s.sweep(inputs, cluster.sweepFeeRate, currentHeight)
+		// errAllSets records the error from broadcasting the sweeping
+		// transactions for all input sets.
+		var errAllSets error
+
+		// allSets contains retried inputs and new inputs. To avoid
+		// creating an RBF for the new inputs, we'd sweep this set
+		// first.
+		for _, inputs := range allSets {
+			errAllSets = s.sweep(
```
Comment (ziggie1984) [inline sweep/sweeper.go:708]: We only return an error when the publishing of the transaction failed if its neither a RBF-Failure nor a DoubleSpend. So I think this will not allow us to broadcast the Newset when RBF conditions are not met because we map a Replacement Err to a [doubleSpend](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/btcwallet/btcwallet.go#L1211-L1216) hence [circumventing](https://github.com/lightningnetwork/lnd/blob/master/sweep/sweeper.go#L1359-L1361) the error.

Maybe as a short term fix before a more sophisticated sweeper, we always return an error when the transaction fails due to publishing and also increase the reattempt times for every input. I don't see any benefit treating unexpected errors of the publisher differently tbh.

**Review by ziggie1984 — CHANGES_REQUESTED** [review]

File: sweep/sweeper.go (hunk 24 above)
Comment (yyforyongyu) [reply sweep/sweeper.go:708]: yeah right, forgot about that part! Maybe I should add an itest to validate the changes, and yes, think we should just return the error here.

**Review by ziggie1984 — APPROVED** [review]
LGTM, one thing I want to highlight, we should definitely add a follow-up PR where we trigger the sweeper when the sweep has the force flag set. Because we have the variable batchwindow-duration people can mess-up there config and end up not sweeping time sensitive inputs.

**Comment by Roasbeef** [comment]
Ultimately, the RBF rules are pretty finicky, and the only way we can ensure we're about to publish a replaceable transaction, is by using `testmempoolaccept` each time. So I consider this a stop gap to start to refractor, simplify the code, and fix any low hanging fruits before we start to integrate something like that.

File: sweep/sweeper.go [hunk 25]
```diff
@@ -1124,38 +1124,39 @@ func (s *UtxoSweeper) sweep(inputs inputSet, feeRate chainfee.SatPerKWeight,
...
-		}),
-	)
-
+	// Publish the sweeping tx with customized label.
 	err = s.cfg.Wallet.PublishTransaction(
 		tx, labels.MakeLabel(labels.LabelTypeSweepTransaction, nil),
 	)
-
-	// In case of an unexpected error, don't try to recover.
-	if err != nil && err != lnwallet.ErrDoubleSpend {
```
Comment (Roasbeef) [inline sweep/sweeper.go:1359]: Why doesn't this try to explicitly catch the double spend error any longer?

Comment (Roasbeef) [reply sweep/sweeper.go:1359]: Is the rationale that since we reschedule everything above, we don't need to specially handle this control flow?

**Review by Roasbeef — APPROVED** [review]
LGTM 🛩️

File: sweep/sweeper.go (hunk 25 above)
Comment (ziggie1984) [reply sweep/sweeper.go:1359]: Exactly we want to make sure, that even `unexpected errors` increase their `attempt counter` allowing them to be removed after the default of 10 attempts (default value). Moreover we use the error to signal, that we try to publish all the new inputs in a separate transaction when the combined one (combination of retried and new inputs) failed.
